src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs

```diff
@@ -603,15 +603,32 @@ internal PragmaWarningState GetPragmaDirectiveWarningState(string id, int positi
             if (_lazyNullableDirectiveMap == null)
             {
                 // Create the #nullable directive map on demand.
-                Interlocked.CompareExchange(ref _lazyNullableDirectiveMap, NullableDirectiveMap.Create(this), null);
+                Interlocked.CompareExchange(ref _lazyNullableDirectiveMap, NullableDirectiveMap.Create(this, IsGeneratedCode()), null);
             }
 
             return _lazyNullableDirectiveMap.GetDirectiveState(position);
         }
 
+        internal bool IsGeneratedCode()
+        {
+            if (_lazyIsGeneratedCode == -1)
+            {
+                // Create the generated code status on demand
+                bool isGenerated = GeneratedCodeUtilities.IsGeneratedCode(
+                           this,
+                           isComment: trivia => trivia.Kind() == SyntaxKind.SingleLineCommentTrivia || trivia.Kind() == SyntaxKind.MultiLineCommentTrivia,
+                           cancellationToken: default);
+
+                Interlocked.CompareExchange(ref _lazyIsGeneratedCode, isGenerated ? 1 : 0, -1);
+            }
+
+            return _lazyIsGeneratedCode == 1;
+        }
+
         private CSharpLineDirectiveMap _lazyLineDirectiveMap;
         private CSharpPragmaWarningStateMap _lazyPragmaWarningStateMap;
         private NullableDirectiveMap _lazyNullableDirectiveMap;
+        private int _lazyIsGeneratedCode = -1;
```

Comment:
>int [](start = 16, length = 3)

Perhaps use `ThreeState`.

Reply:
Sorry, I meant the `enum ThreeState` rather than `bool?`. `bool?` would require an atomic update.

I think either a `ThreeState` value or an `int` as you had originally would be fine. Sorry for the confusion.

---
In reply to: [254506689](https://github.com/dotnet/roslyn/pull/33048#discussion_r254506689) [](ancestors = 254506689)

Reply:
Ok, thanks, have reverted and created #33207 to track fixing this at a later date.